### PR TITLE
Mesh generator cleanup

### PR DESCRIPTION
**Pull request by permcody**
Moving towards a single and consistent way to build the right kind of Mesh between the Mesh and MeshGenerator systems. This PR changes the GeneratedMeshGenerator object to new API in MooseMesh to construct the right kind of MeshBase object.

TODO: We need to evaluate individual MeshGenerators to decide which ones need to be updated. Actually, all should be updated to use the new API, but some might need to force a particular Mesh type.

### CODE REVIEW

File: framework/src/meshgenerators/GeneratedMeshGenerator.C [hunk 1]
```diff
@@ -102,7 +102,8 @@ GeneratedMeshGenerator::GeneratedMeshGenerator(const InputParameters & parameter
 std::unique_ptr<MeshBase>
 GeneratedMeshGenerator::generate()
 {
-  std::unique_ptr<ReplicatedMesh> mesh = libmesh_make_unique<ReplicatedMesh>(comm(), 2);
+  // Have MOOSE construct the correct libMesh::Mesh object using Mesh block and CLI parameters.
+  auto mesh = _mesh->constructMeshBase();
```
Comment (friedmud) [inline framework/src/meshgenerators/GeneratedMeshGenerator.C]: Ok - maybe... but I think `constructMesh()` really needs to take parameters about what's acceptable to be constructed.

**Review by friedmud — CHANGES_REQUESTED** [review]
This is a good start - it could probably go in as it is... but I wouldn't mind if the bit about only constructing the requested kind of mesh made it in here.

Feel free to override if you'd rather do that separately.

**Comment by friedmud** [comment]
Also - I was thinking that the function name should be a bit more clear: `createMeshObject()` or something like that?  Just to differentiate it from `buildMesh()`.

But: I did think about this last night and I think it's a good idea.  It's a nice blend of the old system and the new... and will enable the Generators to pull lots of common data / input from the Mesh block.  I'm sure that will be abused too... but what can we do?

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -2003,9 +1955,81 @@ MooseMesh::clone() const
...
+    {
+      _partitioner_name = "parmetis";
+      _partitioner_overridden = true;
+    }
+  }
+  else
+  {
+    if (override_type == ParallelType::DISTRIBUTED)
+      mooseError("The requested override_type of \"Distributed\" may not be used when MOOSE is "
+                 "running with a DistributedMesh");
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]:
```suggestion
                 "running with a ReplicatedMesh");
```

File: framework/src/mesh/MooseMesh.C [hunk 3]
```diff
@@ -2003,9 +1955,81 @@ MooseMesh::clone() const
...
+    _use_distributed_mesh = true;
+
+  unsigned dim = getParam<MooseEnum>("dim");
+
+  std::unique_ptr<MeshBase> mesh;
+  if (_use_distributed_mesh)
+  {
+    if (override_type == ParallelType::REPLICATED)
+      mooseError("The requested override_type of \"Replicated\" may not be used when MOOSE is "
+                 "running with a DistributedMesh");
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: 👍

File: framework/src/meshgenerators/GeneratedMeshGenerator.C [hunk 4]
```diff
@@ -102,7 +102,8 @@ GeneratedMeshGenerator::GeneratedMeshGenerator(const InputParameters & parameter
 std::unique_ptr<MeshBase>
 GeneratedMeshGenerator::generate()
 {
-  std::unique_ptr<ReplicatedMesh> mesh = libmesh_make_unique<ReplicatedMesh>(comm(), 2);
+  // Have MOOSE construct the correct libMesh::Mesh object using Mesh block and CLI parameters.
+  auto mesh = _mesh->buildMeshBaseObject();
```
Comment (friedmud) [inline framework/src/meshgenerators/GeneratedMeshGenerator.C]: 👍

**Review by friedmud — CHANGES_REQUESTED** [review]
Just one small typo fix

**Comment by permcody** [comment]
Cool - That suggested change thingy actually works with CIVET. This should be good now.

**Review by friedmud — APPROVED** [review]

**Comment by permcody** [comment]
Yeah - except automerge doesn't work with allowed to fail.  That might be something that we look into now that we have this "next" branch setup. If you see it finished testing, otherwise, I'll perform the merge.

**Comment by YaqiWang** [comment]
@permcody Looks like this PR made way to moose master branch and got pushed to Rattlesnake devel. But it is causing one distributed mesh test failing, can you check out https://moosebuild.inl.gov/job/156863/ and propose a fix? Thanks.

**Comment by permcody** [comment]
I assume that the problem is here:
https://github.com/idaholab/moose/commit/5d74fe86fa61da7e93cd1cdf2db1f02b0d576ed7#diff-9f8c79ed34bec90d2c7e006fcd37eeaaR236

Note how the MooseMesh is now created _before_ "check_copy_nodal_vars" action. Considering the name of the failing test, it seems like something probably isn't getting initialized as it probably was before this change? Any idea where to look?

Just so you are clear, what happened with this change is that we are now building the MooseMesh object much earlier, **BUT** we aren't building underlying MeshBase object. The MeshBase object is being built a little bit later now. It might be a few more hours, or early tomorrow before I can take a look. If you have any ideas on where to start, please let me know.

**Comment by YaqiWang** [comment]
That test is using a custom action `AddRestartedVariables`, which is registered to "check_copy_nodal_vars", "add_face_variable", "add_user_object", "add_function" and "add_ic", to facilitate loading variables from exodus file. The action is quite simple and located under `rattlesnake/src/actions/AddRestartedVariables.C`. I suppose it is related with this https://hpcgitlab.hpc.inl.gov/idaholab/rattlesnake/blob/devel/src/actions/AddRestartedVariables.C#L106-113.  I am also busy in something else currently. It will take me several hours to finish...

**Comment by permcody** [comment]
Hmm... No, That action is still after the MeshBase has been constructed and set. I don't immediately see anything in that code that would have been broken.

**Comment by YaqiWang** [comment]
Note this is related only with distributed mesh. Is it possible the run went through a different path?

**Pull request merged**